pageserver/src/tenant/checks.rs

```diff
@@ -44,12 +50,34 @@ pub fn check_valid_layermap(metadata: &[LayerName]) -> Option<String> {
         let lsn_range = layer.lsn_range.clone();
         let intersects = lsn_split_point.range(lsn_range).collect_vec();
         if intersects.len() > 1 {
-            let err = format!(
-                "layer violates the layer map LSN split assumption: layer {} intersects with LSN [{}]",
-                layer,
-                intersects.into_iter().map(|lsn| lsn.to_string()).join(", ")
-            );
-            return Some(err);
+            return check_valid_layermap_slowpath(metadata);
+        }
```

Comment:
If it were a BTreeMap instead of BTreeSet, we wouldn't have to do an `O(n^2)` on all the layers in `metadata` , but could instead just run the slow path against the `.values()` of the `lsn_range`.
No?
Much more efficient.

Reply:
resolved in https://github.com/neondatabase/neon/pull/10044/commits/a16a91f1348fbac77bd5f4d730e4e5119936fefb